sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala

```diff
@@ -142,12 +143,14 @@ case class MapGroups(
  */
 case class CoGroup(
     func: (Any, Iterator[Any], Iterator[Any]) => TraversableOnce[Any],
-    keyObject: Expression,
-    leftObject: Expression,
-    rightObject: Expression,
+    keyDeserializer: Expression,
+    leftDeserializer: Expression,
+    rightDeserializer: Expression,
     serializer: Seq[NamedExpression],
     leftGroup: Seq[Attribute],
     rightGroup: Seq[Attribute],
+    leftAttr: Seq[Attribute],
+    rightAttr: Seq[Attribute],
```

Comment:
yea there is only one place where we create a `CoGroup`, but we do pass `GroupedDataset.dataAttributes` instead of left/right output attributes, https://github.com/apache/spark/pull/10852/files#diff-ed737c05b572cbec98ae5264d2c799b4R314